src/com/esotericsoftware/kryo/KryoBuilder.java

```diff
@@ -0,0 +1,362 @@
+package com.esotericsoftware.kryo;
+
+import java.util.Arrays;
+
+import org.objenesis.strategy.InstantiatorStrategy;
+
+import com.esotericsoftware.kryo.factories.SerializerFactory;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import com.esotericsoftware.kryo.pool.KryoFactory;
+import com.esotericsoftware.kryo.util.DefaultClassResolver;
+import com.esotericsoftware.kryo.util.DefaultStreamFactory;
+import com.esotericsoftware.kryo.util.MapReferenceResolver;
+
+public final class KryoBuilder implements KryoFactory {
+
+	private static enum Step {
+		NEW_INSTANCE,
+		SET_DEFAULT_SERIALIZER,
+		ADD_DEFAULT_SERIALIZER,
+		REGISTER,
+		SET_CLASSLOADER,
+		SET_REGISTRATION_REQUIRED,
+		SET_REFERENCES,
+		SET_COPY_REFERENCES,
+		SET_REFERENCE_RESOLVER,
+		SET_INSTANTIATOR_STRATEGY,
+		SET_AUTO_RESET,
+		SET_MAX_DEPTH,
+		SET_STREAM_FACTORY,
+		SET_ASM_ENABLED,
+		
+		;
+		
+		private static final Step[] VALUES = values();
+	}
+	
+	public static final class KryoBuilderSerializer extends Serializer<KryoBuilder> {
+
+		@SuppressWarnings("synthetic-access")
+		@Override
+		public void write (Kryo kryo, Output output, KryoBuilder object) {
+			output.writeInt(object.step.ordinal(), true);
+			output.writeInt(object.args.length, true);
+			for(int i = 0; i < object.args.length; i++)
+				kryo.writeClassAndObject(output, object.args[i]);
+			kryo.writeObjectOrNull(output, object.tail, KryoBuilder.class);
+		}
+
+		@SuppressWarnings("synthetic-access")
+		@Override
+		public KryoBuilder read (Kryo kryo, Input input, Class<KryoBuilder> type) {
+			KryoBuilder object = new KryoBuilder(null, null, null, null);
+			kryo.reference(object);
+			
+			object.step = Step.VALUES[input.readInt(true)];
+			object.args = new Object[input.readInt(true)];
+			for(int i = 0; i < object.args.length; i++)
+				object.args[i] = kryo.readClassAndObject(input);
+			object.tail = kryo.readObjectOrNull(input, KryoBuilder.class);
+
+			if(object.step == Step.NEW_INSTANCE) {
+				for(KryoBuilder h = object; h != null; h = h.tail)
+					h.head = object;
+			}
+			
+			return object;
+		}
+		
+	}
+	
+	private KryoBuilder head;
+	private KryoBuilder tail;
+	private Step step;
+	private Object[] args;
+
```

Comment:
i think they can be made final, as long as KryoBuilderSerializer never needs to make backreferences; kryo.reference(object) currently happens before deserializing the tail step, and changing this would require removing the call to kryo.reference